doc: update hld for device passthrough (#3868)
projectacrn/acrn-hypervisor: doc/developer-guides/hld/hv-dev-passthrough.rst

Passthrough in the hypervisor provides the following functionalities to
allow VM to access PCI devices directly:

> -  VT-d DMA Remapping for PCI device: hypervisor will setup DMA

Reviewer: for PCI devices?

   remapping during VM initialization phase.

> -  VT-d Interrupt-rempapping for PCI device: hypervisor will enable

Reviewer: ditto

   Passthrough Device Status

Owner of Passthrough Devices
****************************

ACRN hypervisor will do PCI enumeration to discover the PCI devices on the platform.

> According to the hypervisor/VM configurations, the owner of these PCI devices can

Reviewer 2: I think you're missing part of the sentence here?

Author: can "be one of " the ...
will fix the missing word

the following 4 cases:

> - **Hypervisor**: hypervisor uses UART device as the console in debug version for

Reviewer 2: Is this the case for the UART only, or are you using the UART as an example of such device that is assigned and visible only to the hypervisor?

Author: currently is the only one, I will add a description.

  debug purpose, so the UART device is owned by hypervisor and is not visible
  to any VM.
- **Pre-launched VM**: The passthrough devices will be used in a pre-launched VM is

>   pre-defined in vm configuration. These passthrough devices are owned by the

Reviewer 2: ```suggestion
  pre-defined in VM configuration. These passthrough devices are owned by the
```

  pre-launched VM after   the VM is created. These devices will not be removed

>   from the pre-launched VM. There will be pre-launched VM(s) in logical partition

Reviewer 2: ```suggestion
  from the pre-launched VM. There could be pre-launched VM(s) in logical partition
```
?

  mode and hybrid mode.

> - **SOS VM**: All the passthrough devices except these described above (owned by

Reviewer 2: ```suggestion
- **Service VM**: All the passthrough devices except these described above (owned by
```

>   hypervisor or pre-launched VM(s)) are assigned to SOS VM. And some of these devices

Reviewer 2: ```suggestion
  hypervisor or pre-launched VM(s)) are assigned to Service VM. And some of these devices
```

  can be assigned to a post-launched VM according to the passthrough device list
  specified in the parameters of the ACRN DM.
- **Post-launched VM**: A list of passthrough devices can be specified in the parameters of
  the ACRN DM. When creating a post-launched VM, these specified devices will be moved

>   from SOS VM domain to the post-launched VM domain. After the post-launched VM is

Reviewer 2: ```suggestion
  from Service VM domain to the post-launched VM domain. After the post-launched VM is
```

>   powered-off, these devices will be moved back to SOS VM domain.

Reviewer 2: ```suggestion
  powered-off, these devices will be moved back to Service VM domain.
```

   Passthrough Device Status

Owner of Passthrough Devices
****************************

ACRN hypervisor will do PCI enumeration to discover the PCI devices on the platform.
According to the hypervisor/VM configurations, the owner of these PCI devices can
the following 4 cases:
- Hpyervisor: hypervisor uses UART device as the console in debug version for
  debug purpose, so the UART device is ownered by hypervisor and is not visible
  to any VM.
- Pre-launched VM: The passthrough devices will be used in a pre-launched VM is
  pre-defined in vm configuration. These passthrough devices are owned by the
  pre-launched VM after   the VM is created. These devices will not be removed
  from the pre-launched VM. There will be pre-launched VM(s) in logical partition
  mode and hybrid mode.
- SOS VM: All the passthrough devices except these described above (owned by
  hypervisor or pre-launched VM(s)) are assigned to SOS VM. And some of these devices
  can be assigned to a post-launched VM according to the passthrough device list
  specified in the parameters of the ACRN DM.

> - Post-launched VM: A list of passthrough devics can be specified in the paramters of

Reviewer: typo “devics”

ACRN hypervisor checks DMA-Remapping Hardware unit Definition (DRHD) in
host DMAR ACPI table to get basic info, then sets up each DMAR unit. For
simplicity, ACRN reuses EPT table as the translation table in DMAR

> unit for each passthrough device. The control flow of assign and de-assign

Reviewer: assigning and de-assgining?

   ptdev de-assignment control flow

VT-d Interrupt-remapping
************************

The VT-d interrupt-remapping architecture enables system software to

> control and censor external interrupt requests generated by all sources

Reviewer: generated by -> from?

Author: This sentence is from VT-d spec.  Will leave it as it is.

   ptdev de-assignment control flow

VT-d Interrupt-remapping
************************

The VT-d interrupt-remapping architecture enables system software to
control and censor external interrupt requests generated by all sources
including those from interrupt controllers (I/OxAPICs), MSI/MSI-X capable
devices including endpoints, root-ports and Root-Complex integrated
end-points.

> ACRN forces to enabled VT-d interrupt-remapping feature for security reasons.

Reviewer 2: @dbkinder , shouldn't this be rephrased a bit? Maybe something like "ACRN enforces that the VT-d interrupt-remapping feature be enabled for security reasons"

  this, device model is linked with lib pci access to access physical PCI
  device.

MSI-X table emulation
*********************

> VM accesses to MSI-X table should be traped so that hypervisor have the

Reviewer: typo: trapped

...hypervisor has the...

information to map the virtual vector and physcial vector. EPT mapping should
be skipped for the 4KB pages having MSI-X table.
There are three situations for the emulation of MSI-X table:
- SOS: acesses to MSI-X table are handled by HV MMIO handler (4k adjusted up

>   and down). HV will remaps interrupts.

Reviewer: will remap

- Post-launched VM: accesses to MSI-X Tables are handled by DM MMIO handler
  (4k adjusted up and down) and When DM (SOS) writes to the table, it will be

>   intercepted by HV MMIO handler and HV will remaps interrupts.

Reviewer: ditto

  this, device model is linked with lib pci access to access physical PCI
  device.

MSI-X table emulation
*********************

VM accesses to MSI-X table should be trapped so that hypervisor has the
information to map the virtual vector and physical vector. EPT mapping should
be skipped for the 4KB pages having MSI-X table.

There are three situations for the emulation of MSI-X table:

> - **SOS**: accesses to MSI-X table are handled by HV MMIO handler (4k adjusted up

Reviewer 2: ```suggestion
- **Service VM**: accesses to MSI-X table are handled by HV MMIO handler (4K adjusted up
```

  and down). HV will remap interrupts.
- **Post-launched VM**: accesses to MSI-X Tables are handled by DM MMIO handler

>   (4k adjusted up and down) and When DM (SOS) writes to the table, it will be

Reviewer 2: ```suggestion
  (4K adjusted up and down) and when DM (Service VM) writes to the table, it will be
```

  intercepted by HV MMIO handler and HV will remap interrupts.
- **Pre-launched VM**: Writes to MMIO region in MSI-X Table BAR handled by HV MMIO

>   handler. If the offset falls within the MSI-X table [offset, offset+tables_size),

Reviewer 2: ```suggestion
  handler. If the offset falls within the MSI-X table (offset, offset+tables_size),
```

   Initialization of remapping of virtual MSI for SOS

This figure  illustrates how mappings of MSI or MSIX are set up for

> SOS. SOS is responsible for issuing a hypercall to notify the

Reviewer 2: ```suggestion
Service VM. Service VM is responsible for issuing a hypercall to notify the
```

initialization. The device model will create such device entry based on
the physical one according to vendor ID and device ID. Virtualization is
implemented in SOS device model and not in scope of the hypervisor.
For pre-launched VM, ACRN hypervisor doesn't support the ACPI virtualization,
so devices relying on ACPI table are not supported.

GSI Sharing Violation Check
***************************

All the PCI devices that are sharing the same GSI should be assigned to
the same VM to avoid physical GSI sharing between multiple VMs.
In logical partition mode or hybride mode, the PCI devices assigned to
pre-launched VM is statically pre-defined. Developers should take care not to

> voilate the rule.

Reviewer: typo: violate

initialization. The device model will create such device entry based on
the physical one according to vendor ID and device ID. Virtualization is
implemented in SOS device model and not in scope of the hypervisor.
For pre-launched VM, ACRN hypervisor doesn't support the ACPI virtualization,
so devices relying on ACPI table are not supported.

GSI Sharing Violation Check
***************************

All the PCI devices that are sharing the same GSI should be assigned to
the same VM to avoid physical GSI sharing between multiple VMs.
In logical partition mode or hybrid mode, the PCI devices assigned to
pre-launched VM is statically pre-defined. Developers should take care not to
violate the rule.
For post-launched VM, devices that don't support MSI, ACRN DM puts the devices
sharing the same GSI pin to a GSI
sharing group. The devices in the same group should be assigned together to
the current VM, otherwise, none of them should be assigned to the
current VM. A device that violates the rule will be rejected to be

> passed-through. The checking logic is implemented in Device Mode and not

Reviewer 2: ```suggestion
passed-through. The checking logic is implemented in Device Model and not
```
?
